### PR TITLE
Python3.8 support

### PR DESCRIPTION
**Pull request by sklam**
Remaining problems:

- tuple hash algo changed. https://bugs.python.org/issue34751
- looplifting is broken due to CFG assumption changed.

*Edit* The fixes for the above unresolved py3.8 problems will be splitted to separate PRs.  This PR now contains the necessary changes to make Py3.8 mostly working.

### CODE REVIEW

**Comment by sklam** [comment]
Yeah! First CI all pass

**Comment by sklam** [comment]
New code path running on py3.7 now

**Comment by sklam** [comment]
Added py3.8 CI and skip test for looplifting on py3.8

File: numba/byteflow.py [hunk 1]
```diff
@@ -0,0 +1,1139 @@
...
+        # Complete controlflow
+        self._build_cfg(runner.finished)
+        # Prune redundant PHI-nodes
+        self._prune_phis(runner)
+        # Post process
+        for state in sorted(runner.finished, key=lambda x: x.pc_initial):
+            self.block_infos[state.pc_initial] = si = adapt_state_infos(state)
+            _logger.debug("block_infos %s:\n%s", state, si)
+
+        assert self.block_infos
```
Comment (stuartarchibald) [inline numba/byteflow.py]: Given this will propagate to user land, perhaps check and raise a Numba exception instead?

File: numba/byteflow.py [hunk 2]
```diff
@@ -0,0 +1,1139 @@
...
+        # Find use-defs
+        def find_use_defs():
+            defmap = {}
+            phismap = defaultdict(set)
+            for state in runner.finished:
+                for phi, rhs in state._outgoing_phis.items():
+                    if rhs not in phi_set:
+                        # Is a definition
+                        defmap[phi] = state
+                    # else:
```
Comment (stuartarchibald) [inline numba/byteflow.py]: this isn't an else?! the def and phi map both get updated if the phi is a defn ?

File: numba/byteflow.py [hunk 3]
```diff
@@ -0,0 +1,1139 @@
...
+                    phismap[phi].add((rhs, state))
+            _logger.debug("defmap: %s", _lazy_pformat(defmap))
+            _logger.debug("phismap: %s", _lazy_pformat(phismap))
+            return defmap, phismap
+
+        def propagate_phi_map(phismap):
```
Comment (stuartarchibald) [inline numba/byteflow.py]: docs string/comment about what's going on?

File: numba/byteflow.py [hunk 4]
```diff
@@ -0,0 +1,1139 @@
...
+        self._blocks = {}
+        for offset, blockinfo in flow.block_infos.items():
+            self._blocks[offset] = AdaptCFBlock(blockinfo, offset)
+        backbone = self._flow.cfgraph.backbone()
+
+        graph = flow.cfgraph
+        # Find backbone
+        backbone = graph.backbone()
+        # Filter out in loop blocks (Assuming no other cyclic control blocks)
+        # This is to unavoid variable defined in loops to be considered as
```
Comment (stuartarchibald) [inline numba/byteflow.py]: Perhaps `This is to avoid variables defined in loops being considered as function scope`?

**Comment by sklam** [comment]
> Was wondering RE your comment above if the hash tests for tuple are disabled on 3.8, if they are I missed it?

The test are missed due to the test slicing.  `numba.tests.test_hashing.TestTupleHashing` shoud fail on py3.8

File: numba/bytecode.py [hunk 5]
```diff
@@ -206,9 +228,12 @@ def _compute_lineno(cls, table, code):
...
         for offset, lineno in dis.findlinestarts(code):
-            if offset in table:
-                table[offset].lineno = lineno
-        known = -1
+            adj_offset = offset + _FIXED_OFFSET
+            if adj_offset in table:
+                table[adj_offset].lineno = lineno
+        # Assign unfilled lineno
+        # Start with first bytecode's lineno
+        known = table[_FIXED_OFFSET].lineno
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/bytecode.py]: It's being tested already. https://github.com/numba/numba/blob/e4674eea90b451ae2141a20546727bef4ac8b352/numba/tests/test_exceptions.py#L117-L118 is checking line numbers.

If you change the `lineno` in `bytecode.py`, there will be test failures in `numba.tests.test_exceptions`.

File: numba/byteflow.py [hunk 6]
```diff
@@ -0,0 +1,1139 @@
...
+    def op_BUILD_SLICE(self, state, inst):
+        """
+        slice(TOS1, TOS) or slice(TOS2, TOS1, TOS)
+        """
+        argc = inst.arg
+        if argc == 2:
+            tos = state.pop()
+            tos1 = state.pop()
+            start = tos1
+            stop = tos
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/byteflow.py]: This part is mindless copying from dataflow.py https://github.com/numba/numba/blob/e4674eea90b451ae2141a20546727bef4ac8b352/numba/dataflow.py#L611-L623
I also find this easier to correspond to the docstring "slice(TOS2, TOS1, TOS)" given the stack pop order is backwards.

File: numba/byteflow.py [hunk 7]
```diff
@@ -0,0 +1,1139 @@
...
+
+    def op_POP_FINALLY(self, state, inst):
+        # we don't emulate the exact stack behavior
+        if inst.arg != 0:
+            msg = ('Unsupported use of bytecode related to try..finally'
+                   ' or a with-context')
+            raise UnsupportedError(msg, loc=self.get_debug_loc(inst.lineno))
+
+    def op_WITH_CLEANUP_START(self, state, inst):
+        # we don't emulate the exact stack behavior
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/byteflow.py]: More changes coming in the work on supporting try..except. I want to leave it as is until then.

File: numba/byteflow.py [hunk 8]
```diff
@@ -0,0 +1,1139 @@
...
+        count = inst.arg
+        items = list(reversed([state.pop() for _ in range(count)]))
+        lst = state.make_temp()
+        state.append(inst, items=items, res=lst)
+        state.push(lst)
+
+    def op_LIST_APPEND(self, state, inst):
+        value = state.pop()
+        # Python 2.7+ added an argument to LIST_APPEND.
+        if PYVERSION == (2, 6):
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/byteflow.py]: there's a lot of mindless copying from old files, which is only slightly different. i will adjust this.

File: numba/bytecode.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/bytecode.py]: Ah yes, thanks!

File: numba/byteflow.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/byteflow.py]: ok, am happy either way.

File: numba/byteflow.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/byteflow.py]: ok, am happy to leave this until then

File: numba/byteflow.py (hunk 1 above)
Comment (sklam) [reply numba/byteflow.py]: this can actually be removed. added to debug.

File: numba/byteflow.py (hunk 2 above)
Comment (sklam) [reply numba/byteflow.py]: can be safely removed

File: numba/byteflow.py (hunk 3 above)
Comment (sklam) [reply numba/byteflow.py]: added docstring

File: numba/byteflow.py [hunk 9]
```diff
@@ -0,0 +1,1139 @@
...
+        return (self._pc_initial, self._nstack_initial)
+
+    def __hash__(self):
+        return hash(self.get_identity())
+
+    def __eq__(self, other):
+        return self.get_identity() == other.get_identity()
+
+    @property
+    def pc_initial(self):
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/byteflow.py]: Done

**Comment by sklam** [comment]
kicked CI

**Comment by stuartarchibald** [comment]
kick CI again

File: buildscripts/incremental/setup_conda_environment.sh [hunk 10]
```diff
@@ -48,12 +48,18 @@ source activate $CONDA_ENV
...
+else
+    $PIP_INSTALL gitpython
+fi
 
 # Install optional packages into activated env
 if [ "${VANILLA_INSTALL}" != "yes" ]; then
     # Scipy, CFFI, jinja2, IPython and pygments are optional dependencies, but exercised in the test suite
-    $CONDA_INSTALL ${EXTRA_CHANNELS} cffi scipy jinja2 ipython pygments
+    if [ $PYTHON \< "3.8" ]; then
+        $CONDA_INSTALL ${EXTRA_CHANNELS} cffi scipy jinja2 ipython pygments
```
[the remark this answers is not in the document]
Comment (seibert) [reply buildscripts/incremental/setup_conda_environment.sh]: I think we have to revert this one change now that Python 3.8 packages are in numba/main.

File: buildscripts/incremental/setup_conda_environment.sh [hunk 11]
```diff
@@ -70,7 +76,15 @@ elif  [[ $(uname) == Darwin ]]; then
 fi
 
 # Install latest llvmlite build
-$CONDA_INSTALL -c numba llvmlite
+if [ $PYTHON \< "3.8" ]; then
+    $CONDA_INSTALL -c numba llvmlite
+else
+    $CONDA_INSTALL -c numba/label/dev_airflow llvmlite
```
Comment (seibert) [inline buildscripts/incremental/setup_conda_environment.sh]: this should be reverted back to `-c numba` now

Comment (sklam) [reply buildscripts/incremental/setup_conda_environment.sh]: fixed. hopes this is the last change

**Comment by seibert** [comment]
alrighty, going to merge

**Comment by stuartarchibald** [comment]
Good job @sklam :)

**Pull request merged**